lib/trino-parquet/src/main/java/io/trino/parquet/reader/PrimitiveColumnReader.java

```diff
@@ -309,10 +417,30 @@ private ValuesReader initDataReader(ParquetEncoding dataEncoding, int valueCount
 
         try {
             valuesReader.initFromPage(valueCount, in);
+            if (firstRowIndex.isPresent()) {
+                currentRow = firstRowIndex.getAsLong();
+                targetRow = indexIter.hasNext() ? indexIter.peek() : Long.MAX_VALUE;
+            }
             return valuesReader;
         }
         catch (IOException e) {
             throw new ParquetDecodingException("Error reading parquet page in column " + columnDescriptor, e);
         }
     }
+
+    private boolean nextRow(int rl)
+    {
+        if (indexIter == null) {
+            return false;
+        }
+
+        if (rl == 0) {
+            if (currentRow > targetRow) {
+                targetRow = indexIter.hasNext() ? indexIter.next() : Long.MAX_VALUE;
+            }
+            return currentRow++ < targetRow;
```

Comment:
Separate the `++` from the comparison for clarity. It's harder to reason about side effects and ordering of operations the way it's written.